server/src/main/java/org/elasticsearch/index/mapper/StandardIdFieldMapper.java

```diff
@@ -0,0 +1,272 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.index.mapper;
+
+import org.apache.lucene.document.FieldType;
+import org.apache.lucene.index.IndexOptions;
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.search.MatchAllDocsQuery;
+import org.apache.lucene.search.Query;
+import org.apache.lucene.search.SortField;
+import org.apache.lucene.search.TermInSetQuery;
+import org.apache.lucene.util.BytesRef;
+import org.elasticsearch.common.logging.DeprecationCategory;
+import org.elasticsearch.common.logging.DeprecationLogger;
+import org.elasticsearch.common.lucene.Lucene;
+import org.elasticsearch.common.util.BigArrays;
+import org.elasticsearch.index.fielddata.FieldData;
+import org.elasticsearch.index.fielddata.IndexFieldData;
+import org.elasticsearch.index.fielddata.IndexFieldData.XFieldComparatorSource.Nested;
+import org.elasticsearch.index.fielddata.IndexFieldDataCache;
+import org.elasticsearch.index.fielddata.LeafFieldData;
+import org.elasticsearch.index.fielddata.ScriptDocValues;
+import org.elasticsearch.index.fielddata.SortedBinaryDocValues;
+import org.elasticsearch.index.fielddata.fieldcomparator.BytesRefFieldComparatorSource;
+import org.elasticsearch.index.fielddata.plain.PagedBytesIndexFieldData;
+import org.elasticsearch.index.query.SearchExecutionContext;
+import org.elasticsearch.indices.IndicesService;
+import org.elasticsearch.indices.breaker.CircuitBreakerService;
+import org.elasticsearch.script.field.DelegateDocValuesField;
+import org.elasticsearch.script.field.DocValuesField;
+import org.elasticsearch.search.DocValueFormat;
+import org.elasticsearch.search.MultiValueMode;
+import org.elasticsearch.search.aggregations.support.CoreValuesSourceType;
+import org.elasticsearch.search.aggregations.support.ValuesSourceType;
+import org.elasticsearch.search.lookup.SearchLookup;
+import org.elasticsearch.search.sort.BucketedSort;
+import org.elasticsearch.search.sort.SortOrder;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.function.BooleanSupplier;
+import java.util.function.Supplier;
+
+/**
+ * A mapper for the _id field. It does nothing since _id is neither indexed nor
+ * stored, but we need to keep it so that its FieldType can be used to generate
+ * queries.
+ */
+public class StandardIdFieldMapper extends IdFieldMapper {
```

Comment:
Maybe we should call it NoopIdFieldMapper to better reflect what it does (nothing) instead of where it is currently used (standard index mode).

Reply:
It exposes the `_id` for fetching and querying and sometimes provides field data for it. It isn't really noop.

Reply:
`NonGeneratingIdFieldMapper`? 😄

Reply:
Now you are trolling! But, yeah, I don't like the name either. I'm saving this for a follow up.

Reply:
Sorry, I wasn't trying to troll, I am just trying to brainstorm here. How about `ExistingIdFieldMapper` , `ProvidedIdFieldMapper` or `SuppliedIdFieldMapper`?

Reply:
I was sort of trolling. I know it's just brainstorming. `ProvidedIdFieldMapper` seems better. But, like, has the fielddata stuff too.

Reply:
The big difference is that it expects the `_id` to be generated by the coordinating node *or* handed off to us. And it can make fielddata.

Reply:
We went with `ProvidedIdFieldMapper`. It isn't the best name, but it's better than "standard". It gives more of a hint about what makes it unique. Naming is hard.